pkg/cmd/cli/cmd/startbuild.go

```diff
@@ -474,36 +475,78 @@ func streamPathToBuild(git git.Repository, in io.Reader, out io.Writer, client o
 				glog.V(6).Infof("Unable to read Git info from %q: %v", clean, gitErr)
 			}
 
+			// In case we create a separate directory for transferring a Git repo,
+			// we need to have a channel ot make sure we only remove it when the
+			// transfer is finished
+			transferFinished := make(chan bool)
+
 			if asRepo {
 				fmt.Fprintf(out, "Uploading %q at commit %q as binary input for the build ...\n", clean, commit)
 				if gitErr != nil {
 					return nil, fmt.Errorf("the directory %q is not a valid Git repository: %v", clean, gitErr)
 				}
-				pr, pw := io.Pipe()
+
+				// Create a temp directory to move the repo contents to
+				tempDirectory, err := ioutil.TempDir(os.TempDir(), "oc_cloning_"+options.Commit)
+				if err != nil {
+					return nil, err
+				}
+
 				go func() {
-					if err := git.Archive(clean, options.Commit, "tar.gz", pw); err != nil {
-						pw.CloseWithError(fmt.Errorf("unable to create Git archive of %q for build: %v", clean, err))
-					} else {
-						pw.CloseWithError(io.EOF)
+					timeout := 300 * time.Second
+					select {
+					case <-transferFinished:
+					case <-time.After(timeout):
+						glog.Warning("Transferring directory %q timed out after %q seconds.", clean, timeout.Seconds())
 					}
+					// We remove the copy of the directory we made either after the
+					// transfer is finished or 300 seconds, whichever comes first
+					os.RemoveAll(tempDirectory)
 				}()
-				r = pr
+
+				// We only want to grab the contents of the specified commit, with
+				// submodules included
+				cloneOptions := git.CloneOptions{
+					Recursive: true,
+					Shallow:   false,
+					Quiet:     true,
+				}
+
+				if verbose := glog.V(3); !verbose {
+					cloneOptions.Quiet = false
+				}
+
+				// Clone the repository to a temp directory for future tar-ing
+				if err := repo.CloneWithOptions(tempDirectory, path, cloneOptions); err != nil {
+					return nil, err
+				}
+				if err := repo.Checkout(tempDirectory, commit); err != nil {
+					return nil, err
+				}
+				if err := os.RemoveAll(filepath.Join(tempDirectory, ".git")); err != nil {
+					return nil, err
+				}
+
+				// We'll continue to use tar on the cloned directory
+				path = tempDirectory
 
 			} else {
 				fmt.Fprintf(out, "Uploading directory %q as binary input for the build ...\n", clean)
-
-				pr, pw := io.Pipe()
-				go func() {
-					w := gzip.NewWriter(pw)
-					if err := tar.New().CreateTarStream(path, false, w); err != nil {
-						pw.CloseWithError(err)
-					} else {
-						w.Close()
-						pw.CloseWithError(io.EOF)
-					}
-				}()
-				r = pr
 			}
+
+			pr, pw := io.Pipe()
+			go func() {
+				w := gzip.NewWriter(pw)
+				if err := tar.New().CreateTarStream(path, false, w); err != nil {
+					pw.CloseWithError(err)
+				} else {
+					w.Close()
+					pw.CloseWithError(io.EOF)
+				}
+				transferFinished <- false
```

Comment:
I could change it to true, but the only effect that I wanted was to send a signal through to indicating that a deletion is OK.  I don't think we should do error handling on the other side of the channel, because it is specifically for garbage collection.